// fsh-sushi/dist/fshtypes/Invariant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Invariant = void 0;
const _1 = require(".");
const os_1 = require("os");
const common_1 = require("./common");
/**
 * The Invariant class is used to represent the "constraint" field on ElementDefinition
 * Invariant fields map to their corresponding field on "constraint" except:
 * description -> constraint.human
 * name -> constraint.key
 * @see {@link https://www.hl7.org/fhir/elementdefinition.html}
 */
class Invariant extends _1.FshEntity {
    constructor(name) {
        super();
        this.name = name;
    }
    get constructorName() {
        return 'Invariant';
    }
    /**
     * Read only property for id that just returns the name of the invariant
     * This was added so that all types that are returned by FSHTank.fish have an id that can be accessed
     */
    get id() {
        return this.name;
    }
    metadataToFSH() {
        const resultLines = [];
        resultLines.push(`Invariant: ${this.name}`);
        if (this.description) {
            // Description can be a multiline string.
            // If it contains newline characters, treat it as a multiline string.
            if (this.description.indexOf('\n') > -1) {
                resultLines.push(`Description: """${this.description}"""`);
            }
            else {
                resultLines.push(`Description: "${common_1.fshifyString(this.description)}"`);
            }
        }
        if (this.severity) {
            resultLines.push(`Severity: ${this.severity}`);
        }
        if (this.expression) {
            resultLines.push(`Expression: "${common_1.fshifyString(this.expression)}"`);
        }
        if (this.xpath) {
            resultLines.push(`XPath: "${common_1.fshifyString(this.xpath)}"`);
        }
        return resultLines.join(os_1.EOL);
    }
    toFSH() {
        return this.metadataToFSH();
    }
}
exports.Invariant = Invariant;
