// fsh-sushi/dist/fshtypes/ParamRuleSet.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParamRuleSet = void 0;
const escapeRegExp_1 = __importDefault(require("lodash/escapeRegExp"));
const _1 = require(".");
class ParamRuleSet extends _1.FshEntity {
    constructor(name) {
        super();
        this.name = name;
        this.parameters = [];
        this.contents = '';
    }
    applyParameters(values) {
        const paramUsage = new RegExp(`{\\s*(${this.parameters.map(escapeRegExp_1.default).join('|')})\\s*}`, 'g');
        return this.contents.replace(paramUsage, (_fullMatch, paramName) => {
            const matchIndex = this.parameters.indexOf(paramName);
            if (matchIndex > -1) {
                return values[matchIndex];
            }
            else {
                return '';
            }
        });
    }
    getUnusedParameters() {
        return this.parameters.filter(param => {
            return !new RegExp(`{\\s*${escapeRegExp_1.default(param)}\\s*}`).test(this.contents);
        });
    }
}
exports.ParamRuleSet = ParamRuleSet;
