// gofsh/dist/utils/MasterFisher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MasterFisher = void 0;
const fsh_sushi_1 = require("fsh-sushi");
/**
 * The MasterFisher can fish from the LakeOfFHIR and external definitions. When the MasterFisher fishes,
 * it fishes in the LakeOfFHIR first and then the external definitions.  This essentially prefers local
 * definitions first (when there are naming clashes) - matching the SUSHI MasterFisher behavior.
 */
class MasterFisher {
    constructor(lakeOfFHIR, external) {
        this.lakeOfFHIR = lakeOfFHIR;
        this.external = external;
    }
    fishForStructureDefinition(item) {
        const json = this.fishForFHIR(item, fsh_sushi_1.utils.Type.Resource, fsh_sushi_1.utils.Type.Profile, fsh_sushi_1.utils.Type.Extension, fsh_sushi_1.utils.Type.Type);
        if (json) {
            // It is possible we can't parse the json, most likely if it doesn't have a snapshot
            // if that is the case we don't want actual errors, just return nothing
            try {
                return fsh_sushi_1.fhirtypes.StructureDefinition.fromJSON(json);
            }
            catch (_a) { }
        }
    }
    fishForFHIR(item, ...types) {
        var _a, _b;
        return ((_a = this.lakeOfFHIR.fishForFHIR(item, ...types)) !== null && _a !== void 0 ? _a : (_b = this.external) === null || _b === void 0 ? void 0 : _b.fishForFHIR(item, ...types));
    }
    fishForMetadata(item, ...types) {
        var _a, _b;
        return ((_a = this.lakeOfFHIR.fishForMetadata(item, ...types)) !== null && _a !== void 0 ? _a : (_b = this.external) === null || _b === void 0 ? void 0 : _b.fishForMetadata(item, ...types));
    }
    fishForMetadatas(item, ...types) {
        var _a, _b;
        return [
            ...this.lakeOfFHIR.fishForMetadatas(item, ...types),
            ...((_b = (_a = this.external) === null || _a === void 0 ? void 0 : _a.fishForMetadatas(item, ...types)) !== null && _b !== void 0 ? _b : [])
        ];
    }
}
exports.MasterFisher = MasterFisher;
