// diff2html-cli/lib/configuration.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseArgv = void 0;
var path_1 = __importDefault(require("path"));
var types_1 = require("diff2html/lib/types");
function parseArgv(argv) {
    var diff2htmlOptions = {
        outputFormat: argv.style === 'side'
            ? types_1.OutputFormatType.SIDE_BY_SIDE
            : argv.style === 'line'
                ? types_1.OutputFormatType.LINE_BY_LINE
                : undefined,
        diffStyle: argv.diffStyle === 'char' ? types_1.DiffStyleType.CHAR : argv.diffStyle === 'word' ? types_1.DiffStyleType.WORD : undefined,
        matching: argv.matching === 'lines'
            ? types_1.LineMatchingType.LINES
            : argv.matching === 'words'
                ? types_1.LineMatchingType.WORDS
                : argv.matching === 'none'
                    ? types_1.LineMatchingType.NONE
                    : undefined,
        drawFileList: argv.summary !== 'hidden',
        matchWordsThreshold: argv.matchWordsThreshold,
        matchingMaxComparisons: argv.matchingMaxComparisons,
    };
    var defaultWrapperTemplate = path_1.default.resolve(__dirname, '..', 'template.html');
    var configuration = {
        showFilesOpen: argv.summary === 'open' || false,
        synchronisedScroll: argv.synchronisedScroll,
        highlightCode: argv.highlightCode,
        formatType: argv.format,
        outputDestinationType: argv.output,
        outputDestinationFile: argv.file,
        inputSource: argv.input,
        diffyType: argv.diffy,
        htmlWrapperTemplate: argv.htmlWrapperTemplate || defaultWrapperTemplate,
        ignore: argv.ignore || [],
    };
    return [diff2htmlOptions, configuration];
}
exports.parseArgv = parseArgv;
