// diff2html-cli/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.replaceExactly = exports.execute = exports.writeFile = exports.readStdin = exports.readFile = exports.exists = void 0;
var child_process_1 = __importDefault(require("child_process"));
var fs_1 = __importDefault(require("fs"));
function exists(filePath) {
    try {
        return fs_1.default.existsSync(filePath);
    }
    catch (ignore) {
        return false;
    }
}
exports.exists = exists;
function readFile(filePath) {
    return fs_1.default.readFileSync(filePath, 'utf8');
}
exports.readFile = readFile;
function readStdin() {
    return new Promise(function (resolve) {
        var content = '';
        process.stdin.resume();
        process.stdin.on('data', function (buf) {
            content += buf.toString('utf8');
        });
        process.stdin.on('end', function () { return resolve(content); });
    });
}
exports.readStdin = readStdin;
function writeFile(filePath, content) {
    return fs_1.default.writeFileSync(filePath, content);
}
exports.writeFile = writeFile;
function execute(executable, args) {
    return child_process_1.default.spawnSync(executable, args, { encoding: 'utf8' }).stdout;
}
exports.execute = execute;
function replaceExactly(value, searchValue, replaceValue) {
    return value.replace(searchValue, function () { return replaceValue; });
}
exports.replaceExactly = replaceExactly;
