// fsh-sushi/dist/fhirtypes/CodeSystem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeSystem = void 0;
const lodash_1 = require("lodash");
const mixins_1 = require("./mixins");
const Mixin_1 = require("../utils/Mixin");
/**
 * Class representing a FHIR R4 CodeSystem
 * @see {@link https://www.hl7.org/fhir/codesystem.html}
 */
class CodeSystem {
    constructor() {
        this.resourceType = 'CodeSystem';
        this.status = 'active';
        this.content = 'complete';
    }
    /**
     * Get the file name for serializing to disk.
     * @returns {string} the filename
     */
    getFileName() {
        return `CodeSystem-${this.id}.json`;
    }
    /**
     * Exports the CodeSystem to a properly formatted FHIR JSON representation.
     * @returns {any} the FHIR JSON representation of the CodeSystem
     */
    toJSON() {
        return Object.assign({}, lodash_1.cloneDeep(this));
    }
}
exports.CodeSystem = CodeSystem;
Mixin_1.applyMixins(CodeSystem, [mixins_1.HasId, mixins_1.HasName]);
