// fsh-sushi/dist/import/loadConfigurationFromIgResource.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const ini_1 = __importDefault(require("ini"));
const fhir_1 = require("fhir/fhir");
const utils_1 = require("../utils");
/**
 * Attempts to find an ig resource file and extract the required configuration properties from it
 * @param input - path to the input directory
 * @returns {Configuration} the extracted configuration
 */
function loadConfigurationFromIgResource(input) {
    var _a, _b;
    // First, look in the ig.ini file for a path to the IG resource
    let igPath;
    const igIniPath = path_1.default.join(input, '..', 'ig.ini');
    if (fs_extra_1.default.existsSync(igIniPath)) {
        try {
            const igIni = ini_1.default.parse(fs_extra_1.default.readFileSync(igIniPath, 'utf-8'));
            igPath = path_1.default.join(input, '..', igIni.ig);
        }
        catch (_c) { }
    }
    // Make a list of possible path, if ig.ini exists and points to a ig file, add just that
    // otherwise consider all files in the input folder of the ig
    const igInputPath = path_1.default.join(input, '..', 'input');
    const possibleIgPaths = [];
    if (fs_extra_1.default.existsSync(igPath)) {
        possibleIgPaths.push(igPath);
    }
    else if (fs_extra_1.default.existsSync(igInputPath)) {
        possibleIgPaths.push(...fs_extra_1.default.readdirSync(igInputPath).map(file => path_1.default.join(igInputPath, file)));
    }
    // Go through each possible path, and check each xml/json file to see if they are an IG resource
    const fhirConverter = new fhir_1.Fhir();
    let igResource;
    let multipleIgs = false;
    possibleIgPaths.forEach(filePath => {
        let fileContent;
        if (path_1.default.extname(filePath) === '.json') {
            fileContent = fs_extra_1.default.readJSONSync(filePath);
        }
        else if (path_1.default.extname(filePath) === '.xml') {
            fileContent = fhirConverter.xmlToObj(fs_extra_1.default.readFileSync(filePath).toString());
        }
        if ((fileContent === null || fileContent === void 0 ? void 0 : fileContent.resourceType) === 'ImplementationGuide') {
            // If 2 possible IG resources are found, we cannot tell which to use, so return
            if (igResource != null) {
                multipleIgs = true;
                utils_1.logger.error('Multiple possible ImplementationGuide resources in "input" folder, so no configuration can be extracted.' +
                    ' Ensure only one ImplementationGuide resource is in the "input" folder, or give the path to the desired resource in ig.ini.');
            }
            igResource = fileContent;
            igPath = filePath;
        }
    });
    // Extract the configuration from the resource
    if (igResource && igResource.url && !multipleIgs) {
        utils_1.logger.info(`Extracting FSHOnly configuration from ${igPath}.`);
        return {
            canonical: igResource.url.replace(/\/ImplementationGuide.*/, ''),
            FSHOnly: true,
            dependencies: (_a = igResource.dependsOn) === null || _a === void 0 ? void 0 : _a.filter(dep => dep.packageId && dep.version),
            fhirVersion: (_b = igResource.fhirVersion) !== null && _b !== void 0 ? _b : [],
            version: igResource.version
        };
    }
    return null;
}
exports.loadConfigurationFromIgResource = loadConfigurationFromIgResource;
